uploader/push_to_sling.py

```diff
@@ -0,0 +1,98 @@
+  placeholderArt.ignore.jpg'
+ 
+''')
+parser.add_argument('--server', '-s', help='The Pantheon server to upload modules to, default http://localhost:8080', default='http://localhost:8080')
+parser.add_argument('--repository', '-r', help='The name of the repository; this is a convenience for the user. If not specified, one will be autogenerated', default=getpass.getuser() + '_' + datetime.now().strftime('%Y-%m-%d_%H-%M-%S'))
+parser.add_argument('--user', '-u', help='Username for authentication, default \'demo\'', default='demo')
+parser.add_argument('--password', '-p', help='Password for authentication, default \'demo\'. If \'-\' is supplied, the script will prompt for the password.', default='demo')
+parser.add_argument('--directory', '-d', help='Directory to upload, default is current working directory. (' + os.getcwd() + ')', default=os.getcwd())
+args = parser.parse_args()
+
+pw = args.password
+if pw == '-':
+    pw = getpass.getpass()
+
+print('Using server: ' + args.server)
+print('Using repository: ' + args.repository)
+print('--------------')
+
+directory_in_str = args.directory
+pathlist = Path(directory_in_str).glob('**/*.*')
+
+for path in pathlist:
+    # relative path to the current directory (with extension)
+    relative_path_in_str = str(path.relative_to(Path.cwd()))
+    # parent directory
+    parent_dir_str = str(path.parent.relative_to(Path.cwd()))
+    if parent_dir_str == '.':
+        parent_dir_str = ''
+    # file becomes a/file/name (no extension)
+    base_name = path.stem
+
+    url = args.server + "/content/repositories/" + args.repository
+    if parent_dir_str:
+        url += '/' + parent_dir_str
+
+    # print('base name: ' + base_name)
+
+    # Asciidoc content (treat as a module)
+    if path.suffix == '.adoc' or path.suffix == '.asciidoc':
+        print(path)
+        if base_name.endswith(('.module', '.internal', '.title')):  # Should differentiate later
+            url += '/' + path.name
+            # print(url)
+            data = {"jcr:primaryType": 'pant:module',
+                    "jcr:title": base_name,
+                    "jcr:description": base_name,
+                    "sling:resourceType": 'pantheon/modules',
+                    "pant:originalName": path.name,
+                    "asciidoc@TypeHint": 'nt:file'}
+            files = {'asciidoc': ('asciidoc', open(path, 'rb'), 'text/x-asciidoc')}
+
+            # Minor question: which is correct, text/asciidoc or text/x-asciidoc?
+            # It is text/x-asciidoc. Here's why:
+            # https://tools.ietf.org/html/rfc2045#section-6.3
+            # Paraphrased: "If it's not an IANA standard, use the 'x-' prefix.
+            # Here's the list of standards; text/asciidoc isn't in it.
+            # https://www.iana.org/assignments/media-types/media-types.xhtml#text
+
+            r = requests.post(url, headers=HEADERS, data=data, files=files, auth=(args.user, pw))
```

Comment:
It's good practice to wrap most http calls with some exception handling, since response codes are in the very high numbers and growing, so some of the more critical ones can be handled or ignored. Let's discuss this with Ankit about what we want our python guidelines to be. But we need to handle when 1 out of 100 files comes back with a network error or 400 response. Should the rest be deleted? Marked incomplete? All uploaded in a state where nobody has viewing rights to them until all arrive as a cohesive unit? Let's brainstorm one morning about the outcomes of various use cases.

Reply:
Good thinking, Ivan. I'll leave this as-is for now, but you're absolutely right, we should implement some form of error-handling as a future story.